idb/postgres/internal/writer/writer.go

```diff
@@ -139,47 +139,52 @@ func setSpecialAccounts(addresses transactions.SpecialAddresses, batch *pgx.Batc
 
 // Get the ID of the creatable referenced in the given transaction
 // (0 if not an asset or app transaction).
-func transactionAssetID(block *bookkeeping.Block, intra uint64, typeenum idb.TxnTypeEnum) uint64 {
+func transactionAssetID(txn transactions.SignedTxnWithAD, intra uint64, block *bookkeeping.Block) uint64 {
 	assetid := uint64(0)
-	stxnad := block.Payset[intra]
-	txn := block.Payset[intra].Txn
 
-	switch typeenum {
-	case idb.TypeEnumAssetConfig:
+	switch txn.Txn.Type {
+	case protocol.ApplicationCallTx:
+		assetid = uint64(txn.ApplicationID)
+		if assetid == 0 {
+			assetid = uint64(txn.ApplyData.ApplicationID)
+		}
+		if assetid == 0 {
+			// pre v30 transactions do not have ApplyData.ConfigAsset or InnerTxns
+			// so txn counter + payset pos calculation is OK
+			assetid = block.TxnCounter - uint64(len(block.Payset)) + intra + 1
+		}
+	case protocol.AssetConfigTx:
 		assetid = uint64(txn.ConfigAsset)
 		if assetid == 0 {
-			if stxnad.ApplyData.ConfigAsset != 0 {
-				assetid = uint64(stxnad.ApplyData.ConfigAsset)
-			} else {
-				// pre v30 transactions do not have ApplyData.ConfigAsset or InnerTxns
-				// so txn counter + payset pos calculation is OK
-				assetid = block.TxnCounter - uint64(len(block.Payset)) + intra + 1
-			}
+			assetid = uint64(txn.ApplyData.ConfigAsset)
 		}
-	case idb.TypeEnumAssetTransfer:
-		assetid = uint64(txn.XferAsset)
-	case idb.TypeEnumAssetFreeze:
-		assetid = uint64(txn.FreezeAsset)
-	case idb.TypeEnumApplication:
-		assetid = uint64(txn.ApplicationID)
 		if assetid == 0 {
-			if stxnad.ApplyData.ApplicationID != 0 {
-				assetid = uint64(stxnad.ApplyData.ApplicationID)
-			} else {
-				// pre v30 transactions do not have ApplyData.ApplicationID or InnerTxns
-				// so txn counter + payset pos calculation is OK
-				assetid = block.TxnCounter - uint64(len(block.Payset)) + intra + 1
-			}
+			// pre v30 transactions do not have ApplyData.ApplicationID or InnerTxns
+			// so txn counter + payset pos calculation is OK
+			assetid = block.TxnCounter - uint64(len(block.Payset)) + intra + 1
 		}
+	case protocol.AssetTransferTx:
+		assetid = uint64(txn.Txn.XferAsset)
+	case protocol.AssetFreezeTx:
+		assetid = uint64(txn.Txn.FreezeAsset)
 	}
 
 	return assetid
 }
 
+func countInner(stxnad transactions.SignedTxnWithAD) uint64 {
```

Comment:
The function name is misleading since we also count the outer transaction. Perhaps rename it to `countTxns`?